### ISSUE TITLE
WebDAV Android SSL handshake error (self-signed)

### ISSUE THREAD

**Issue by Tursu311** [super-productivity/super-productivity#2855]
### Your Environment

- Version used: In-App: 7.13.2 APK, F-Droid version: 21.0 17/04/2023
- Operating System and version: Android 12 - LineageOS 19.1 MicroG version

### Expected Behavior

When I put my server info for sync using WebDAV same as in .AppImage version, should sync my tasks with Nextcloud WebDAV directly as in desktop, maybe showing a pop-up to confirm SSL certificate is self-signed

### Current Behavior

When I save my config, and press sync button, upper left screen appears a popup lower part of the screen: Unknown Sync Error: Error: Invalid response: -2 SL Handshake Error. It never syncs my tasks.

### Steps to Reproduce (for bugs)

1. Use android app 
2. Input WebDAV sync config for a Nextcloud self-signed instance
3. Save and press sync.

### Can you reproduce this reliably?

Every time, tried different passwords, folders, and users for WebDAV.

### Console Output

No idea to check on android

**Comment by gostaj** [comment]
Syncing using WebDAV against a serving having a self-signed certificate works fine with the desktop app on a Mac.
But I'm having this problem myself in the Android app and would like to sync with my server having a self-signed certificate.
But looking at the code it looks "as designed".

@johannesjo if I would create a PR for accepting invalid certificates in the Android app, would that be of interest?
I'm using [Joplin](https://github.com/laurent22/joplin) for note syncing and it has a toggle for "Ignore TLS certificate errors", I'm thinking of something similar.
Or the Android app should work like the desktop app of Super Productivity that accepts self-signed certificates by default.

**Comment by Jagdfalke** [comment]
> Syncing using WebDAV against a server using a self-signed certificate works fine with the desktop app on a Mac. But I'm having this problem myself in the Android app and would like to sync with my server having a self-signed certificate. But looking at the code it looks "as designed" that it doesn't work in the Android app.
> 
> @johannesjo if I would create a PR for accepting invalid certificates in the Android app, would that be of interest? I'm using [Joplin](https://github.com/laurent22/joplin) for note syncing and it has a toggle for "Ignore TLS certificate errors", I'm thinking of something similar. Or the Android app should work like the desktop app of Super Productivity that accepts self-signed certificates by default.

PRs are always welcome! What you described is a very viable solution. I'm all for it.

**Comment by johannesjo** [comment]
There is nothing similar implemented for WebDAV. Not a big fan of this workaround neither tbh, as it might confuse users that don't need it and opens up security vulnerabilities.

**Comment by gostaj** [comment]
I understand your concern @johannesjo but correct me if I'm wrong: The Super Productivity Mac OS app currently accepts self-signed https certificates, by default, over WebDAV sync without any errors.
(The Mac OS app successfully syncs against my WebDAV server that is using a self-signed certificate.)
But this is not possible today using the Super Productivity Android app where the "SSL handshake error" is seen.

My PR (if you want it) would be adding an option to the Android app, turned off by default, that accepts self-signed https certificates over WebDAV sync similar to the [Joplin app](https://github.com/laurent22/joplin):
<img width="222" alt="Screenshot 2024-12-05 at 21 55 01" src="https://github.com/user-attachments/assets/13528411-ddaa-4317-8b9a-f3a21ee2f82f">

**Comment by johannesjo** [comment]
Can you maybe explain your use case to me? Since let's encrypt exists, it should not be difficult to get a valid certificate.

**Comment by gostaj** [comment]
> Can you maybe explain your use case to me? Since let's encrypt exists, it should not be difficult to get a valid certificate.

Yes, I have a WebDAV account on a server with a self-signed certificate. Since I'm not the admin I cannot setup Let's encrypt. Syncing works fine, as I mentioned, using the MacOS app but not with the Android app.
Another option is of course to use HTTP but that is even less secure than HTTPS with a self-signed certificate.

I see your point with confusing users and opening up vulnerabilities. To me this would be an advanced setting for people knowing what they are doing, disabled by default and perhaps use the wording "not recommended".

Personally I love apps that are configurable and let me, the user, be the judge of which ways I want to sync my own data. I mean WebDAV over HTTP is supported and it is even less secure.

Having that said I really like and respect your project and won't bother you if you don't want this feature.

**Comment by johannesjo** [comment]
Thank you very much for the explanation. I think we would have to hide it in some sort of advanced settings section (same for Jira) and even than it is a little bit of extra clutter and code to maintain, so personally I'd lean towards not adding it at the moment. But I might change my mind, if this use case turns out to be more common than I expected :)

Sorry, about this and thank you very much for being so understanding!

**Comment by gostaj** [comment]
Totally understand, no problem.
Perhaps this thread is a bad place to aks but are there any other Android bugs/features you need help with?

**Comment by johannesjo** [comment]
Oh there is a lot of stuff that could and should be improved and help is always welcome. Generally I recommend working on something you care about personally, but there is also a list of issues tagged with android:
https://github.com/johannesjo/super-productivity/issues?q=is%3Aopen+is%3Aissue+label%3Aandroid

If you have some new feature you'd like to see, it's best to open up an issue about it first, so we can discuss the details.

**Comment by KeithBoehler** [comment]
> Can you maybe explain your use case to me? Since let's encrypt exists, it should not be difficult to get a valid certificate.

So I just found this trying to solve the same issue. I just want to add that I am self hosting, but not exposing my server to the web just doing things locally. I am rather new to self hosting. I have been using the ignore ssl on Joplin as well.

**Comment by CBerndt-Work** [comment]
> Can you maybe explain your use case to me? Since let's encrypt exists, it should not be difficult to get a valid certificate.

Hi, I'm in a corporate setting and am having this issue with jira as issue provider and the client on windows. The self-signed certificate in this case is provided by the corporate CA. The root certificate is installed on the system and honored in other applications; not sure where the difference is.
My preferred solution would be the ability to add accepted certificates and maybe proxy settings for the app. This would be more extra code to maintain, but it fixes the problem in a safe way.

**Comment by FairTradeOrange** [comment]
I have the same Problem with a valid LetsEncrypt Certificate?
On Windows no problem to sync, on Android the "SSL Handshake Error".

Windows Version: 11.3
Android Version: 12.0.1_L2A

I use the WebDav Service from my Synology DSM415+.

I like the application really much and use it with Obsidian for my work, university and general life stuff. 
But yes, Android Sync would definitely be great, no matter which method ( Webdav, SFTP, FTPS, SMB, HTTPS,... ). 

Thank you for creating such a nice productivity tool! ( :

**Comment by FairTradeOrange** [comment]
Ok. After a lot of testing, sniffing, monitoring and research, i have finally found the solution for problems with an android application sync through webdav. 

In short, the problem "Network Error" or "SSL Handshake Error" exists, because of missing CA certificates in the android system. 
A valid certificate, in my case from LetsEncrypt, can only be used if the following is done:

1. Get informations about your Intermediate CA
    - for example "Let’s Encrypt R11", you can find informations under "certificate details" on a website in the browser
2. Download the .pem file ( in my case: https://letsencrypt.org/certificates/  )
3. Install the Certficiate .pem file in Android  
   - goto Settings -> Security -> Encryption and Credentials -> Install Certificate -> Install CA

Everything should working now. 
If you use a selfsigned Certificate, this should also work with your own CA cert.
Maybe this "experience" and information can help someone with the same problems.

**Comment by johannesjo** [comment]
Thank you very much @FairTradeOrange !

If this is a common issue(?), we should provide your answer as feedback to the user.

**Comment by khorchanov** [comment]
Hello, 
I just came here to say that I'm a happy user, but when I switched to a lan-hosted version of nextcloud, sup-prod android is not accepting the self signed certificate of my server which is annoying.
I'd be really happy if the app just let advanced users accep a self signed certificate (exactly as the android nextcloud app is doing)

Also, sup-prod desktop seems to not care about the invalid certificates, the android app behavior becomes inconsistent with the desktop app.

I'm a noob, but I'm keen to drop a PR if you give me some guidance.

Thanks a lot

**Comment by Nowaker** [comment]
Any update? We need feature parity between desktop and Android apps. The former allows self-signed or expired certificates, so the latter should too.

**Comment by johannesjo** [comment]
@Nowaker PRs for this are very welcome! I myself currently have a lot of other subjects on the plate.... :/

**Comment by novikov1337danil** [comment]
You probably just need to complete this one step
      
>     3. Install the Certficiate .pem file in Android
>        
>        * goto Settings -> Security -> Encryption and Credentials -> Install Certificate -> Install CA

You don't even need to issue the certificate via letsencrypt (only if you really want it)

@johannesjo I would suggest closing this issue

**Comment by johannesjo** [comment]
@novikov1337danil I would still accept PRs for this, if they would implement this as discussed above. It will be closed automatically if noone volunteers in 180 days.

**Comment by Smoky62** [comment]
Hi. Many thanks for the programm at first. I would like the implementation of beeing able to use self signed certificate too. At the time i have to use http because of non accepting self signed certificates  in android by the app. I've installed the self signed certificate in Android as an app certificate, but nevertheless https syncing fails (i think it only accepts CA certifikates). Otherwise letsencrypt with an self-hosted dns is a little overkill for syncing in an private network imhop.
